File: src/main/java/ru/reosfire/WMLib/Guis/Inventory/Components/GuiComponent.java
package ru.reosfire.WMLib.Guis.Inventory.Components;

import org.bukkit.inventory.Inventory;
import ru.reosfire.WMLib.Guis.Inventory.Gui;
import ru.reosfire.WMLib.Guis.Inventory.IClickEventHandler;
import ru.reosfire.WMLib.Text.Replacement;
import ru.reosfire.WMLib.Yaml.Default.Gui.CommandButtonConfig;
import ru.reosfire.WMLib.Yaml.Default.Gui.ComponentConfig;
import ru.reosfire.WMLib.Yaml.Default.Gui.RawItemConfig;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Set;

public abstract class GuiComponent
{
    private static final HashMap<String, IComponentCreator> creators = new HashMap<>();
    public static void AddCreator(String type, IComponentCreator creator)
    {
        creators.put(type, creator);
    }
    static
    {
        AddCreator("RawItem".toLowerCase(),
                (config, player) -> new RawItem(new RawItemConfig(config.getSection()), player));
        AddCreator("CommandButton".toLowerCase(),
                (config, player) -> new CommandButton(new CommandButtonConfig(config.getSection()), player));
    }

    private final Set<IClickEventHandler> clickEventHandlers = new HashSet<>();
    public static GuiComponent Create(ComponentConfig config, Gui gui)
    {
        return creators.get(config.Type).Create(config, gui);
    }

    protected final Gui gui;

    public GuiComponent(Gui gui)
    {
        this.gui = gui;
    }
    public final void AddComponent(GuiComponent component)
    {
        gui.AddComponent(component);
    }
    public final void ReRender(Replacement... replacements)
    {
        gui.RedrawComponent(this, replacements);
    }

    public abstract void Register();
    public void Unregister()
    {
        for (IClickEventHandler clickEventHandler : clickEventHandlers)
        {
            gui.Events.removeClickHandler(clickEventHandler);
        }
    }

    protected void addClickHandler(IClickEventHandler handler)
    {
        gui.Events.addClickHandler(handler);
        clickEventHandlers.add(handler);
    }

    public abstract void RenderTo(Inventory inventory, Replacement... replacements);
}

File: src/main/java/ru/reosfire/wmlib/guis/inventory/components/GuiComponent.kt
package ru.reosfire.wmlib.guis.inventory.components

import org.bukkit.inventory.Inventory
import ru.reosfire.wmlib.guis.inventory.Gui
import ru.reosfire.wmlib.guis.inventory.IClickEventHandler
import java.util.HashSet
import java.util.HashMap
import java.util.Locale
import ru.reosfire.wmlib.yaml.common.gui.ComponentConfig
import ru.reosfire.wmlib.yaml.common.gui.RawItemConfig
import ru.reosfire.wmlib.text.Replacement
import ru.reosfire.wmlib.yaml.common.gui.CommandButtonConfig

abstract class GuiComponent(protected val gui: Gui) {
    private val clickEventHandlers: MutableSet<IClickEventHandler> = HashSet()
    fun addComponent(component: GuiComponent?) {
        gui.addComponent(component!!)
    }

    fun reRender(vararg replacements: Replacement) {
        gui.redrawComponent(this, *replacements)
    }

    abstract fun register()
    open fun unregister() {
        for (clickEventHandler in clickEventHandlers) {
            gui.Events.removeClickHandler(clickEventHandler)
        }
    }

    protected fun addClickHandler(handler: IClickEventHandler) {
        gui.Events.addClickHandler(handler)
        clickEventHandlers.add(handler)
    }

    abstract fun renderTo(inventory: Inventory, vararg replacements: Replacement)

    companion object {
        private val creators = HashMap<String, IComponentCreator>()
        fun addCreator(type: String, creator: IComponentCreator) {
            creators[type] = creator
        }

        init {
            addCreator("RawItem".lowercase(Locale.getDefault())) { config: ComponentConfig, gui: Gui ->
                RawItem(RawItemConfig(config.section), gui)
            }
            addCreator("CommandButton".lowercase(Locale.getDefault())) { config: ComponentConfig, gui: Gui ->
                CommandButton(CommandButtonConfig(config.section), gui)
            }
        }

        fun create(config: ComponentConfig, gui: Gui): GuiComponent {
            return creators[config.type]!!.create(config, gui)
        }
    }
}